File: 2015/2015-01.py
#!/usr/bin/env python

from typing import LiteralString
import time

Input = list[str] | list[LiteralString]


def part_1(input: Input) -> int:
    return sum(1 if c == "(" else -1 for c in input[0])


def part_2(input: Input):
    s = 0
    for i, c in enumerate(input[0]):
        s += 1 if c == "(" else -1
        if s == -1:
            return i + 1


def _test():
    def assert_eq(a, b):
        assert a == b, f"{a} != {b}"

    assert_eq(part_1([")())())"]), -3)
    assert_eq(part_2(["()())"]), 5)


def run(fn, year=2015, day=1, part=0):
    start = time.perf_counter_ns()
    res = fn()
    elapsed_ns = time.perf_counter_ns() - start
    print(f"{year}\t{day}\t{part}\t{res}\t{elapsed_ns}")
    return res


if __name__ == "__main__":
    with open("2015-01.input") as f:
        input_file = [line.rstrip("\n") for line in f.readlines()]
    _test()
    run(lambda: part_1(input_file), part=1)
    run(lambda: part_2(input_file), part=2)

File: 2015/2015-01.zig
const std = @import("std");

fn part1(input: []const u8) i64 {
    var count: i64 = 0;
    for (input) |c| {
        if (c == '(') {
            count += 1;
        } else if (c == ')') {
            count -= 1;
        }
    }
    return count;
}

fn part2(input: []const u8) usize {
    var count: i64 = 0;
    for (input, 0..) |c, i| {
        if (c == '(') {
            count += 1;
        } else if (c == ')') {
            count -= 1;
        }
        if (count == -1) {
            return i + 1;
        }
    }
    return 0;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    const size = (try std.fs.cwd().statFile("2015-01.input")).size;
    const input = try std.fs.cwd().readFileAlloc(allocator, "2015-01.input", size);

    var timer = try std.time.Timer.start();
    const p1 = part1(input);
    std.debug.print("part 1: {} in {} μs\n", .{ p1, timer.read() / 1000 });
    timer.reset();
    const p2 = part2(input);
    std.debug.print("part 2: {} in {} μs\n", .{ p2, timer.read() / 1000 });
}

test "part 1" {
    try std.testing.expectEqual(part1(")())())"), -3);
}

test "part 2" {
    try std.testing.expectEqual(part2("()())"), 5);
}
